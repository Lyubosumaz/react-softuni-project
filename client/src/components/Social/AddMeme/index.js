import { useState } from 'react';
import { history } from '../../../utils/history';
import { connect } from 'react-redux';
import { httpSocial } from '../../../services/http';
import schema from './add-meme-validations';
import { toastSuccess, toastError } from '../../../utils/toastHandler';
import { componentData } from '../../../class-names.json';

function AddMeme(props) {
    const title = useFormInput('');
    const imageUrl = useFormInput('');
    const [errors, setErrors] = useState({});

    function handleSubmit(e) {
        e.preventDefault();
        const meme = {
            title: title.value,
            imageUrl: imageUrl.value,
        };
        const hasErrors = Object.keys(errors).filter((key) => errors[key].length > 0);

        if (hasErrors.length === 0 && meme.title && meme.imageUrl && props.isLogin) {
            httpSocial
                .addMeme(meme)
                .then((res) => {
                    toastSuccess(res);
                    history.push('/social');
                })
                .catch((err) => {
                    toastError(err);
                });
        }
    }

    function useFormInput(initialValue) {
        const [value, setValue] = useState(initialValue);

        function handleChange(event) {
            setValue(event.target.value);
            validate(event);
        }

        function validate(event) {
            const name = event.target.id;

            schema.fields[name]
                .validate(event.target.value, { abortEarly: false })
                .then(() => {
                    setErrors({ ...errors, [name]: [] });
                })
                .catch((err) => {
                    setErrors({ ...errors, [name]: err.errors });
                });
        }

        return { value, onChange: handleChange };
    }

    return (
        <section className={`${componentData}`}>
            <form>
                <div className="form-div-container">
                    <label htmlFor="Title">
                        <b>Title:</b>
                    </label>
                    <input type="text" placeholder="Write some funny title" className="form-input" id="title" onChange={useFormInput} {...title} />
                    {errors.title && <div className="form-input-error">{errors.title[0]}</div>}
                </div>

                <div className="form-div-container">
                    <label htmlFor="imageUrl">
                        <b>imageUrl:</b>
                    </label>
                    <input type="text" placeholder="Copy and Paste your memeURL here" className="form-input" id="imageUrl" onChange={useFormInput} {...imageUrl} />
                    {errors.imageUrl && <div className="form-input-error">{errors.imageUrl[0]}</div>}
                </div>

                <div>
                    <button type="submit" className="form-action-btn" onClick={handleSubmit}>
                        Submit
                    </button>
                </div>
            </form>
        </section>
    );
}

function mapStateToProps(state) {
    return {
        isLogin: state.user.isLogin,
    };
}

export default connect(mapStateToProps)(AddMeme);
