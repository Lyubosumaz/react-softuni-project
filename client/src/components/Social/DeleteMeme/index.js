import { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { history } from '../../../utils/history';
import { httpSocial } from '../../../services/http';
import { setNotification } from '../../Notification/actions';
import { formComponent, formFieldsWrapper, formBtnClass } from '../../../utils/class-names.json';

function DeleteMeme({ isLogin, memeId, setNotificationSuccess, setNotificationError }) {
    const isLogged = isLogin;
    const memeIdProps = memeId;
    const [meme, setMeme] = useState(null);

    useEffect(() => {
        httpSocial.getMeme(memeIdProps).then((meme) => {
            setMeme(meme);
        });
    }, [memeIdProps]);

    function handleDeleteMeme(e) {
        e.preventDefault();
        if (isLogged) {
            httpSocial
                .deleteMeme(memeIdProps)
                .then((res) => {
                    setNotificationSuccess(res);
                    history.push('/social');
                })
                .catch((err) => {
                    setNotificationError(err);
                });
        }
    }

    return (
        <section className={`${formComponent}`}>
            <div className={`${formFieldsWrapper}`}>
                <form onSubmit={handleDeleteMeme}>
                    <div className="form-field">
                        <label htmlFor="Title">
                            <span>Title:</span>
                        </label>

                        {meme && <input type="text" className="form-input" id="title" value={meme.title} disabled />}
                    </div>

                    <div className="form-field">
                        <label htmlFor="imageUrl">
                            <span>imageUrl:</span>
                        </label>

                        {meme && <input type="text" className="form-input" id="imageUrl" value={meme.imageUrl} disabled />}
                    </div>

                    <div className="form-field-buttons">
                        <input type="reset" className={formBtnClass} value="reset" />
                        <input type="submit" className={formBtnClass} value="delete meme" />
                    </div>
                </form>
            </div>
        </section>
    );
}

function mapStateToProps(state) {
    return {
        isLogin: state.user.isLogin,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        setNotificationSuccess: (data) => dispatch(setNotification().success(data)),
        setNotificationError: (data) => dispatch(setNotification().error(data)),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(DeleteMeme);

DeleteMeme.propTypes = {
    isLogin: PropTypes.bool.isRequired,
    memeId: PropTypes.string.isRequired,
    setNotificationSuccess: PropTypes.func.isRequired,
    setNotificationError: PropTypes.func.isRequired,
};
