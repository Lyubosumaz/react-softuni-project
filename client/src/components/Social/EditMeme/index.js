import { PropTypes } from 'prop-types';
import { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { httpSocial } from '../../../services/http';
import { formBtnClass, formComponent, formFieldsWrapper } from '../../../utils/class-names.json';
import { history } from '../../../utils/history';
import { setNotification } from '../../Notification/actions';
import schema from './edit-meme-validations';

function EditMeme({ isLogin, memeId, setNotificationSuccess, setNotificationError }) {
    const isLogged = isLogin;
    const memeIdProps = memeId;
    const [meme, setMeme] = useState(null);
    const [title, setTitle] = useState('');
    const [imageUrl, setImageUrl] = useState('');
    const [errors, setErrors] = useState({});

    useEffect(() => {
        httpSocial.getMeme(memeIdProps).then((meme) => {
            setMeme(meme);
            setTitle(meme.title);
            setImageUrl(meme.imageUrl);
        });
    }, [memeIdProps]);

    function handleTitle(event) {
        setTitle(event.target.value);
        validate(event);
    }

    function handleInputUrl(event) {
        setImageUrl(event.target.value);
        validate(event);
    }

    function validate(event) {
        const name = event.target.id;

        schema.fields[name]
            .validate(event.target.value, { abortEarly: false })
            .then(() => {
                setErrors({ ...errors, [name]: [] });
            })
            .catch((err) => {
                setErrors({ ...errors, [name]: err.errors });
            });
    }

    function handleEditMeme(e) {
        e.preventDefault();
        const data = {
            id: memeIdProps,
            title,
            imageUrl,
        };
        const hasErrors = Object.keys(errors).filter((key) => errors[key].length > 0);

        if (hasErrors.length === 0 && data.title && data.imageUrl && isLogged) {
            httpSocial
                .editMeme(data)
                .then((res) => {
                    setNotificationSuccess(res);
                    history.push('/social');
                })
                .catch((err) => {
                    setNotificationError(err);
                });
        }
    }

    return (
        <section className={`${formComponent}`}>
            <div className={`${formFieldsWrapper}`}>
                <form onSubmit={handleEditMeme}>
                    <div className="form-field">
                        <label htmlFor="Title">
                            <span>Title:</span>
                        </label>

                        {meme && <input type="text" placeholder="Write some funny title" className="form-input" id="title" onChange={handleTitle} value={title} />}

                        {Array.isArray(errors.title) && errors.title[0] ? (
                            <div className="form-error-message">
                                <span>{errors.title[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field">
                        <label htmlFor="imageUrl">
                            <span>imageUrl:</span>
                        </label>

                        {meme && <input type="text" placeholder="Copy and Paste your memeURL here" className="form-input" id="imageUrl" onChange={handleInputUrl} value={imageUrl} />}

                        {Array.isArray(errors.imageUrl) && errors.imageUrl[0] ? (
                            <div className="form-error-message">
                                <span>{errors.imageUrl[0]}</span>
                            </div>
                        ) : (
                            <div className="form-error-container">
                                <span>Error Container</span>
                            </div>
                        )}
                    </div>

                    <div className="form-field-buttons">
                        <input type="reset" className={formBtnClass} value="reset" />
                        <input type="submit" className={formBtnClass} value="edit meme" />
                    </div>
                </form>
            </div>
        </section>
    );
}

function mapStateToProps(state) {
    return {
        isLogin: state.user.isLogin,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        setNotificationSuccess: (data) => dispatch(setNotification(data).success()),
        setNotificationError: (data) => dispatch(setNotification(data).error()),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(EditMeme);

EditMeme.propTypes = {
    isLogin: PropTypes.bool.isRequired,
    memeId: PropTypes.string.isRequired,
    setNotificationSuccess: PropTypes.func.isRequired,
    setNotificationError: PropTypes.func.isRequired,
};
