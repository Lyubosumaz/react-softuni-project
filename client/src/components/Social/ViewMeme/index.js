import { useState, useEffect } from 'react';
import { PropTypes } from 'prop-types';
import { httpSocial } from '../../../services/http';
import MemeCard from '../../MemeCard';
import { componentData } from '../../../utils/class-names.json';

export default function ViewMeme({ memeId }) {
    const memeIdProps = memeId;
    const [meme, setMeme] = useState(null);

    useEffect(() => {
        httpSocial.getMeme(memeIdProps).then((meme) => {
            setMeme(meme);
        });
    }, [memeIdProps]);

    return <section className={`${componentData}`}>{meme && <MemeCard num={0} meme={meme} />}</section>;
}

ViewMeme.propTypes = {
    memeId: PropTypes.string.isRequired,
};
