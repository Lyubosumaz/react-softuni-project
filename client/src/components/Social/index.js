import { useState, useRef, useCallback } from 'react';
import { handleRoute } from '../../utils/history';
import MemeCard from './components/MemeCard';
import useMemePage from './useMemePage';
import { componentData } from '../../class-names.json';

export default function Social() {
    const [pageNumber, setPageNumber] = useState(1);
    const { memes, loading, error, hasMore } = useMemePage(pageNumber);

    const observer = useRef();
    const lastMemeCardRef = useCallback(
        (node) => {
            if (loading) {
                return;
            }

            if (observer.current) {
                observer.current.disconnect();
            }

            observer.current = new IntersectionObserver((entires) => {
                if (entires[0].isIntersecting && hasMore) {
                    setPageNumber(pageNumber + 1);
                }
            });

            if (node) {
                observer.current.observe(node);
            }
        },
        [loading, hasMore, pageNumber]
    );

    return (
        <section className={`${componentData}`}>
            <div>
                <button className="active-button" onClick={handleRoute('/404')}>
                    Create Meme
                </button>
            </div>
            <div>
                <button className="active-button" onClick={handleRoute('/social/add-meme')}>
                    Add Meme
                </button>
            </div>

            <h1>Memes</h1>

            <div className="memes-container">
                {memes.map((meme, index) => {
                    if (memes.length === index + 1) {
                        return (
                            <div key={meme.index} ref={lastMemeCardRef}>
                                <MemeCard meme={meme} />
                            </div>
                        );
                    } else {
                        return (
                            <div key={meme.index}>
                                <MemeCard meme={meme} />
                            </div>
                        );
                    }
                })}
            </div>
            <div>{loading && 'Loading...'}</div>
            <div>{error && 'Error'}</div>
        </section>
    );
}
