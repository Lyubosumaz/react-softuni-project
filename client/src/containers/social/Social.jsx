import { useState, useRef, useCallback } from 'react';
import handleRoute from '../../utils/handleRoutes';
import MemeCard from './components/meme-card/MemeCard';
import useMemePage from './useMemePage';

export default function Social() {
    const [pageNumber, setPageNumber] = useState(1);
    const { memes, loading, error, hasMore } = useMemePage(pageNumber);

    const observer = useRef();
    const lastMemeCardRef = useCallback(
        (node) => {
            if (loading) {
                return;
            }
            if (observer.current) {
                observer.current.disconnect();
            }
            observer.current = new IntersectionObserver((entires) => {
                if (entires[0].isIntersecting && hasMore) {
                    setPageNumber(pageNumber + 1);
                }
            });
            if (node) {
                observer.current.observe(node);
            }
        },
        [loading, hasMore, pageNumber]
    );

    return (
        <div className="social-container">
            <h1>Social</h1>

            <div>
                <button className="active-button" onClick={handleRoute('/404')}>
                    Create Meme
                </button>
            </div>
            <div>
                <button className="active-button" onClick={handleRoute('/social/add-meme')}>
                    Add Meme
                </button>
            </div>

            <h1>Memes</h1>

            <div className="memes-container">
                {memes.map((meme, index) => {
                    if (memes.length === index + 1) {
                        return (
                            <div key={meme.index} ref={lastMemeCardRef}>
                                <MemeCard meme={meme} />
                            </div>
                        );
                    } else {
                        return (
                            <div key={meme.index}>
                                <MemeCard meme={meme} />
                            </div>
                        );
                    }
                })}
            </div>
            <div>{loading && 'Loading...'}</div>
            <div>{error && 'Error'}</div>

            <div className="info-container">
                <p>
                    Join the race and climb to the top{' '}
                    <button className="info-button" onClick={handleRoute('/games')}>
                        Game
                    </button>
                    !
                </p>
            </div>
        </div>
    );
}
