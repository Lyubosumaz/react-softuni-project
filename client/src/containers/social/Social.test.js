import React from 'react';
import { shallow } from 'enzyme';
import Social from './Social';

it('should renter HouseOfFame Component', () => {
    const wrapper = shallow(<Social />);
    expect(wrapper).toBeTruthy();
});

it('should checks for html changes', () => {
    const wrapper = shallow(<Social />);
    expect(wrapper).toMatchSnapshot();
});
