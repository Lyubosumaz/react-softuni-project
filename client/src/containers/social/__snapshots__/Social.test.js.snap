// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should checks for html changes 1`] = `
<div
  className="main-container"
>
  <h1>
    Social
  </h1>
  <div>
    <button
      className="active-button"
      onClick={[Function]}
    >
      Add Meme
    </button>
  </div>
  <h1>
    Memes
  </h1>
  <div
    className="memes-container"
  />
  <div>
    Loading...
  </div>
  <div />
  <div
    className="info-container"
  >
    <p>
      Join the race and climb to the top 
      <button
        className="info-button"
        onClick={[Function]}
      >
        Game
      </button>
      !
    </p>
  </div>
</div>
`;
